File: apps/api/app/schemas/backup.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel


class SnapshotResponse(BaseModel):
    id: str
    device_id: str
    type: str
    raw_config: str
    checksum_sha: str
    version_label: str
    created_at: str


class ConfigVersionResponse(BaseModel):
    id: str
    device_id: str
    parent_ver_id: Optional[str] = None
    branch: str
    author: str
    checksum_sha: str
    raw_config: str
    change_summary: Optional[str] = None
    risk_level: str
    created_at: str


class RollbackPreviewRequest(BaseModel):
    device_id: str
    current_version_id: str
    target_version_id: str
    selective_mode: str = "FULL"


class SafetyCheckItem(BaseModel):
    name: str
    status: str


class RollbackPreviewResponse(BaseModel):
    device_id: str
    current_version_id: str
    target_version_id: str
    selective_mode: str
    unified_diff: str
    is_safety_validated: bool
    safety_checks: List[SafetyCheckItem]


class CreateChangeRequestPayload(BaseModel):
    title: str
    description: str
    category: str = "STANDARD"
    target_devices: List[str]


class ChangeApprovalStepResponse(BaseModel):
    id: str
    step_name: str
    approver_name: str
    is_approved: bool
    comments: Optional[str] = None


class ChangeRequestResponse(BaseModel):
    id: str
    organization_id: str
    requester_name: str
    title: str
    description: str
    category: str
    status: str
    target_devices: List[str]
    approvals: List[ChangeApprovalStepResponse]
    created_at: str

File: packages/types/src/backup.ts
export enum SnapshotType {
  RUNNING_CONFIG = 'RUNNING_CONFIG',
  STARTUP_CONFIG = 'STARTUP_CONFIG',
  CANDIDATE_CONFIG = 'CANDIDATE_CONFIG',
  GOLDEN_CONFIG = 'GOLDEN_CONFIG',
  LAST_KNOWN_GOOD = 'LAST_KNOWN_GOOD',
  PRE_CHANGE_SNAPSHOT = 'PRE_CHANGE_SNAPSHOT',
  POST_CHANGE_SNAPSHOT = 'POST_CHANGE_SNAPSHOT',
  EMERGENCY_SNAPSHOT = 'EMERGENCY_SNAPSHOT',
}

export enum ChangeCategory {
  EMERGENCY = 'EMERGENCY',
  STANDARD = 'STANDARD',
  MAJOR = 'MAJOR',
  MINOR = 'MINOR',
}

export enum ChangeApprovalStatus {
  DRAFT = 'DRAFT',
  PENDING_REVIEW = 'PENDING_REVIEW',
  PENDING_APPROVAL = 'PENDING_APPROVAL',
  APPROVED = 'APPROVED',
  REJECTED = 'REJECTED',
  EXECUTED = 'EXECUTED',
  CANCELLED = 'CANCELLED',
}

export interface ConfigSnapshotEntity {
  id: string;
  deviceId: string;
  type: SnapshotType;
  rawConfig: string;
  checksumSha: string;
  versionLabel: string;
  createdAt: string;
}

export interface ConfigVersionEntity {
  id: string;
  deviceId: string;
  parentVerId?: string;
  branch: string;
  author: string;
  checksumSha: string;
  rawConfig: string;
  changeSummary?: string;
  riskLevel: string;
  createdAt: string;
}

export interface ChangeApprovalStep {
  id: string;
  stepName: string;
  approverName: string;
  isApproved: boolean;
  comments?: string;
  decidedAt?: string;
}

export interface ChangeRequestEntity {
  id: string;
  organizationId: string;
  requesterName: string;
  title: string;
  description: string;
  category: ChangeCategory;
  status: ChangeApprovalStatus;
  targetDevices: string[];
  approvals: ChangeApprovalStep[];
  createdAt: string;
}

export interface RollbackPreviewResult {
  deviceId: string;
  currentVersionId: string;
  targetVersionId: string;
  selectiveMode: 'FULL' | 'SELECTIVE_LINES' | 'SELECTIVE_INTERFACES';
  unifiedDiff: string;
  isSafetyValidated: boolean;
  safetyChecks: { name: string; status: 'PASSED' | 'FAILED' }[];
}
